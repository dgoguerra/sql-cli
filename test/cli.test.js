const { runCli, getTestKnex, getKnexUri } = require("./utils");

const TEST_DATETIME_1 = "2020-07-24 18:34:00";
const TEST_DATETIME_2 = "2020-07-24 19:25:00";

describe("CLI basic commands", () => {
  let knex;
  let connUri;

  beforeAll(async () => {
    knex = getTestKnex();
    connUri = getKnexUri(knex);

    await knex.schema.createTable("table_1", (t) => {
      t.increments("id");
      t.integer("field_1");
      t.text("field_2");
      t.timestamps();
    });
    await knex.schema.createTable("table_2", (t) => {
      t.increments("id");
      t.decimal("field_1").notNullable();
      t.text("field_2").defaultTo("default text");
      t.timestamps();
    });
    await knex.schema.createTable("table_3", (t) => {
      t.bigIncrements("idField");
      t.bigInteger("field_1");
      t.timestamps();
    });
    await knex("table_1").insert([
      {
        field_1: 12,
        field_2: "foo",
        created_at: TEST_DATETIME_1,
      },
      {
        field_1: 30,
        field_2: "bar",
        created_at: TEST_DATETIME_1,
        updated_at: TEST_DATETIME_1,
      },
    ]);
    await knex("table_2").insert([
      {
        field_1: 12.3,
        field_2: "foo",
        created_at: TEST_DATETIME_1,
        updated_at: TEST_DATETIME_1,
      },
      {
        field_1: 30.45,
        field_2: "bar",
        created_at: TEST_DATETIME_1,
        updated_at: TEST_DATETIME_2,
      },
    ]);
  });

  it("can list tables", async () => {
    expect(await runCli("list", [connUri])).toMatchSnapshot();
  });

  it("can show table", async () => {
    expect(await runCli("show", [`${connUri}/table_2`])).toMatchSnapshot();
  });

  it("can diff tables", async () => {
    expect(
      await runCli("diff", [`${connUri}/table_1`, `${connUri}/table_2`])
    ).toMatchSnapshot();
  });

  it("can diff schemas", async () => {
    expect(await runCli("diff", [connUri, connUri])).toMatchSnapshot();
  });

  it("can create conn aliases", async () => {
    await runCli("alias", [
      "add",
      "alias1",
      "mysql://app:secret@127.0.0.1:1234/dbname",
    ]);
    await runCli("alias", ["add", "alias2", "sqlite3:///path/to/file.db"]);

    // Alias has been created
    expect(await runCli("alias", ["ls"])).toMatchSnapshot();
  });

  it("can use a created alias", async () => {
    await runCli("alias", ["add", "test-alias", connUri]);
    expect(await runCli("list", ["test-alias"])).toMatchSnapshot();
  });

  it("can delete conn aliases", async () => {
    await runCli("alias", ["rm", "test-alias"]);
    await runCli("alias", ["rm", "alias1"]);

    // Only alias2 still exists
    expect(await runCli("alias", ["ls"])).toMatchSnapshot();

    await runCli("alias", ["rm", "alias2"]);
  });
});
