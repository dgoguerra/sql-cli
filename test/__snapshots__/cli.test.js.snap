// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CLI basic tests mssql can create conn aliases 1`] = `
"alias  conn                                     
alias1 mysql://app:secret@127.0.0.1:1234/dbname 
alias2 sqlite3:///path/to/file.db               

"
`;

exports[`CLI basic tests mssql can delete conn aliases 1`] = `
"alias  conn                       
alias2 sqlite3:///path/to/file.db 

"
`;

exports[`CLI basic tests mssql can diff schemas 1`] = `
"Diff of database schemas:

No tables changes

"
`;

exports[`CLI basic tests mssql can diff tables 1`] = `
"Diff of tables schema

column  type                   
field_1 int nullable → decimal 

4x similar, 1x changed

Diff of tables content (first 100 rows)

id field_1    updated_at                                          
1  12 → 12.3  [null] → 2020-07-24T18:34:00.000Z                   
2  30 → 30.45 2020-07-24T18:34:00.000Z → 2020-07-24T19:25:00.000Z 

"
`;

exports[`CLI basic tests mssql can list tables 1`] = `
"table   rows bytes 
table_1 2    16 kB 
table_2 2    16 kB 
table_3 0    0 B   

(32 kB in 3 tables)
"
`;

exports[`CLI basic tests mssql can show table 1`] = `
"column     type         nullable 
id         int          false    
field_1    decimal      false    
field_2    nvarchar(-1) true     
created_at datetime2    true     
updated_at datetime2    true     

"
`;

exports[`CLI basic tests mssql can use a created alias 1`] = `
"table   rows bytes 
table_1 2    16 kB 
table_2 2    16 kB 
table_3 0    0 B   

(32 kB in 3 tables)
"
`;

exports[`CLI basic tests mysql2 can create conn aliases 1`] = `
"alias  conn                                     
alias1 mysql://app:secret@127.0.0.1:1234/dbname 
alias2 sqlite3:///path/to/file.db               

"
`;

exports[`CLI basic tests mysql2 can delete conn aliases 1`] = `
"alias  conn                       
alias2 sqlite3:///path/to/file.db 

"
`;

exports[`CLI basic tests mysql2 can diff schemas 1`] = `
"Diff of database schemas:

No tables changes

"
`;

exports[`CLI basic tests mysql2 can diff tables 1`] = `
"Diff of tables schema

column  type                   
field_1 int nullable → decimal 

4x similar, 1x changed

Diff of tables content (first 100 rows)

id field_1    updated_at                                          
1  12 → 12.30 [null] → 2020-07-24T18:34:00.000Z                   
2  30 → 30.45 2020-07-24T18:34:00.000Z → 2020-07-24T19:25:00.000Z 

"
`;

exports[`CLI basic tests mysql2 can list tables 1`] = `
"table   rows bytes   
table_1 2    16.4 kB 
table_2 2    16.4 kB 
table_3 0    16.4 kB 

(49.2 kB in 3 tables)
"
`;

exports[`CLI basic tests mysql2 can show table 1`] = `
"column     type        nullable 
id         int         false    
field_1    decimal     false    
field_2    text(65535) true     
created_at datetime    true     
updated_at datetime    true     

"
`;

exports[`CLI basic tests mysql2 can use a created alias 1`] = `
"table   rows bytes   
table_1 2    16.4 kB 
table_2 2    16.4 kB 
table_3 0    16.4 kB 

(49.2 kB in 3 tables)
"
`;

exports[`CLI basic tests pg can create conn aliases 1`] = `
"alias  conn                                     
alias1 mysql://app:secret@127.0.0.1:1234/dbname 
alias2 sqlite3:///path/to/file.db               

"
`;

exports[`CLI basic tests pg can delete conn aliases 1`] = `
"alias  conn                       
alias2 sqlite3:///path/to/file.db 

"
`;

exports[`CLI basic tests pg can diff schemas 1`] = `
"Diff of database schemas:

No tables changes

"
`;

exports[`CLI basic tests pg can diff tables 1`] = `
"Diff of tables schema

column  type                       
field_1 integer nullable → numeric 

4x similar, 1x changed

Diff of tables content (first 100 rows)

id field_1    updated_at                                          
1  12 → 12.30 [null] → 2020-07-24T18:34:00.000Z                   
2  30 → 30.45 2020-07-24T18:34:00.000Z → 2020-07-24T19:25:00.000Z 

"
`;

exports[`CLI basic tests pg can list tables 1`] = `
"table   rows bytes 
table_1 0          
table_2 0          
table_3 0          

(0 B in 3 tables)
"
`;

exports[`CLI basic tests pg can show table 1`] = `
"column     type                     nullable 
id         integer                  false    
field_1    numeric                  false    
field_2    text                     true     
created_at timestamp with time zone true     
updated_at timestamp with time zone true     

"
`;

exports[`CLI basic tests pg can use a created alias 1`] = `
"table   rows bytes 
table_1 0          
table_2 0          
table_3 0          

(0 B in 3 tables)
"
`;

exports[`CLI basic tests sqlite3 can create conn aliases 1`] = `
"alias  conn                                     
alias1 mysql://app:secret@127.0.0.1:1234/dbname 
alias2 sqlite3:///path/to/file.db               

"
`;

exports[`CLI basic tests sqlite3 can delete conn aliases 1`] = `
"alias  conn                       
alias2 sqlite3:///path/to/file.db 

"
`;

exports[`CLI basic tests sqlite3 can diff schemas 1`] = `
"Diff of database schemas:

No tables changes

"
`;

exports[`CLI basic tests sqlite3 can diff tables 1`] = `
"Diff of tables schema

column  type                     
field_1 integer nullable → float 

4x similar, 1x changed

Diff of tables content (first 100 rows)

id field_1    updated_at                                
1  12 → 12.3  [null] → 2020-07-24 18:34:00              
2  30 → 30.45 2020-07-24 18:34:00 → 2020-07-24 19:25:00 

"
`;

exports[`CLI basic tests sqlite3 can list tables 1`] = `
"table   rows bytes 
table_1            
table_2            
table_3            

(0 B in 3 tables)
"
`;

exports[`CLI basic tests sqlite3 can show table 1`] = `
"column     type     nullable 
id         integer  false    
field_1    float    false    
field_2    text     true     
created_at datetime true     
updated_at datetime true     

"
`;

exports[`CLI basic tests sqlite3 can use a created alias 1`] = `
"table   rows bytes 
table_1            
table_2            
table_3            

(0 B in 3 tables)
"
`;
